### PR TITLE
Contracts optimization

### PR DESCRIPTION
**Pull request by szotov**
* Size and storage optimization
* `createPolicy` optimization
* Fixes `divideStake`
* Adds methods to check refund

### CODE REVIEW

File: nucypher/blockchain/eth/sol/source/contracts/MinersEscrow.sol [hunk 1]
```diff
@@ -442,13 +465,18 @@ contract MinersEscrow is Issuer {
...
-        confirmActivity(msg.sender, lockedTokens, 0);
+        confirmActivity(msg.sender, lockedTokens, 0, lastActivePeriod);
     }
 
     /**
     * @notice Mint tokens for sender for previous periods if he locked his tokens and confirmed activity
     **/
-    function mint() public onlyTokenOwner {
+    function mint() external onlyTokenOwner {
+        MinerInfo storage info = minerInfo[msg.sender];
```
Comment (jMyles) [inline nucypher/blockchain/eth/sol/source/contracts/MinersEscrow.sol]: What's happening here?

File: nucypher/blockchain/eth/sol/source/contracts/MinersEscrow.sol [hunk 2]
```diff
@@ -519,18 +547,15 @@ contract MinersEscrow is Issuer {
     )
         internal returns (uint256 reward)
     {
-        uint256 amount;
         for (uint256 i = 0; i < _info.stakes.length; i++) {
             StakeInfo storage stake =  _info.stakes[i];
             uint256 lastPeriod = getLastPeriod(stake, _startPeriod);
             if (stake.firstPeriod <= _period && lastPeriod >= _period) {
-                (amount, _info.decimals) = mint(
+                reward = reward.add(mint(
```
Comment (jMyles) [inline nucypher/blockchain/eth/sol/source/contracts/MinersEscrow.sol]: Yeah, this reads more nicely.

File: nucypher/blockchain/eth/sol/source/contracts/Issuer.sol [hunk 3]
```diff
@@ -102,23 +102,22 @@ contract Issuer is Upgradeable {
...
+        uint256 _allLockedPeriods
     )
-        internal returns (uint256 amount, uint256 decimals)
+        internal returns (uint256 amount)
     {
-        // TODO finish method before calculation after end of mining
         uint256 currentSupply = _period <= lastMintedPeriod ?
             Math.min256(currentSupply1, currentSupply2) :
             Math.max256(currentSupply1, currentSupply2);
+        if (currentSupply == futureSupply) {
```
[the remark this answers is not in the document]
Comment (szotov) [reply nucypher/blockchain/eth/sol/source/contracts/Issuer.sol]: `token.totalSupply()` will call the token contract that costs gas.
`totalSupply` suits more than `futureSupply`, I'll rename it

File: nucypher/blockchain/eth/sol/source/contracts/MinersEscrow.sol [hunk 4]
```diff
@@ -64,7 +64,7 @@ contract MinersEscrow is Issuer {
         uint256 lastActivePeriod;
         Downtime[] downtime;
         StakeInfo[] stakes;
-        bytes32[] minerIds;
+        bytes[] minerIds;
```
[the remark this answers is not in the document]
Comment (szotov) [reply nucypher/blockchain/eth/sol/source/contracts/MinersEscrow.sol]: Yes, now it can store any byte array.
I'm not sure about other ways to store keys

**Pull request merged**